crates/hyperqueue/src/worker/streamer.rs

```diff
@@ -1,314 +1,205 @@
     instance_id: InstanceId,
 }
 
 impl StreamSender {
-    pub async fn close(&self) -> tako::Result<()> {
+    pub async fn flush(&self) -> tako::Result<()> {
+        let (sender, receiver) = oneshot::channel();
         if self
             .sender
-            .send(FromStreamerMessage::End(EndTaskStreamMsg {
-                task: self.task_id,
-                instance: self.instance_id,
-            }))
+            .send(StreamerMessage::Flush(sender))
             .await
             .is_err()
         {
-            return Err("Sending close message failed".into());
+            return Err("Sending flush message failed".into());
         }
+        receiver.await.map_err(|_| {
+            tako::Error::GenericError("Stream failed while flushing stream".to_string())
+        })?;
         Ok(())
     }
 
     pub async fn send_data(&self, channel: ChannelId, data: Vec<u8>) -> tako::Result<()> {
         if self
             .sender
-            .send(FromStreamerMessage::Data(DataMsg {
-                task: self.task_id,
-                instance: self.instance_id,
-                channel,
+            .send(StreamerMessage::Write {
+                header: StreamChunkHeader {
+                    time: Utc::now(),
+                    job: self.job_id,
+                    task: self.job_task_id,
+                    instance: self.instance_id,
+                    channel,
+                    size: data.len() as u64,
+                },
                 data,
-            }))
+            })
             .await
             .is_err()
         {
             return Err("Sending streamer message failed".into());
         }
         Ok(())
     }
+}
 
-    pub async fn send_stream_start(&self) -> tako::Result<()> {
-        if self
-            .sender
-            .send(FromStreamerMessage::Start(StartTaskStreamMsg {
-                task: self.task_id,
-                instance: self.instance_id,
-            }))
-            .await
-            .is_err()
-        {
-            return Err("Sending streamer message failed".into());
+async fn stream_writer(
+    streamer_ref: &StreamerRef,
+    path: &Path,
+    mut receiver: Receiver<StreamerMessage>,
+) -> crate::Result<()> {
+    let uid: String = rand::thread_rng()
+        .sample_iter(&Alphanumeric)
+        .take(12)
+        .map(char::from)
+        .collect::<String>();
+    let mut path = path.to_path_buf();
+    path.push(format!("{uid}.hqs"));
+    log::debug!("Opening stream file {}", path.display());
+    let mut file = BufWriter::new(File::create(path).await?);
+    file.write_all(STREAM_FILE_HEADER).await?;
+    let mut buffer: Vec<u8> = Vec::new();
+    {
+        let streamer = streamer_ref.get();
+        let header = StreamFileHeader {
+            server_uid: Cow::Borrowed(&streamer.server_uid),
+            worker_id: streamer.worker_id,
+        };
+        bincode_config().serialize_into(&mut buffer, &header)?;
+    };
+    file.write_all(&buffer).await?;
```

Comment:
Why is this not written to the `file` directly?

Reply:
Because serde does not support direct writing to async files